File: backend/patient-service/app/redis/redis_cache_service.py
import asyncio
import json
import random
import uuid
from collections.abc import Awaitable, Callable
from typing import TypeVar

from app.redis.redis_service import RedisService


T = TypeVar("T")


class RedisCacheService:

    def __init__(self):
        self.redis_service = RedisService()

    async def get_or_set(
        self,
        key: str,
        fetcher: Callable[[], Awaitable[T]],
        ttl: int,
        lock_ttl: int = 10,
        retries: int = 5,
        retry_delay: float = 0.05,
        jitter: int = 30,
    ) -> T:

        # 1. Check cache
        cached = await self.redis_service.get(key)

        if cached:
            return json.loads(cached)

        # 2. Create distributed lock
        lock_key = f"lock:{key}"
        lock_token = str(uuid.uuid4())

        acquired = await self.redis_service.acquire_lock(
            lock_key,
            lock_token,
            lock_ttl,
        )

        if acquired:
            try:
                # 3. Double-check cache
                cached = await self.redis_service.get(key)

                if cached:
                    return json.loads(cached)

                # 4. Fetch from database
                data = await fetcher()

                # 5. Add TTL jitter
                final_ttl = ttl

                if jitter > 0:
                    final_ttl += random.randint(0, jitter)

                # 6. Store in Redis
                await self.redis_service.set(
                    key,
                    json.dumps(data),
                    final_ttl,
                )

                return data

            finally:
                # 7. Release only our lock
                await self.redis_service.release_lock(
                    lock_key,
                    lock_token,
                )

        # 8. Another request owns the lock.
        # Wait for it to populate the cache.
        for _ in range(retries):
            await asyncio.sleep(retry_delay)

            cached = await self.redis_service.get(key)

            if cached:
                return json.loads(cached)

        # 9. Lock holder failed.
        return await fetcher()

File: backend/health-institute-service/src/redis/redis-cache.service.ts
import { Injectable } from '@nestjs/common';
import { randomUUID } from 'crypto';
import { RedisService } from './redis.service';

interface CacheOptions {
  ttl: number;
  lockTtl?: number;
  retries?: number;
  retryDelay?: number;
  jitter?: number;
}

@Injectable()
export class RedisCacheService {
  constructor(private readonly redisService: RedisService) {}

  async getOrSet<T>(
    key: string,
    fetcher: () => Promise<T>,
    options: CacheOptions,
  ): Promise<T> {
    const {
      ttl,
      lockTtl = 10,
      retries = 5,
      retryDelay = 50,
      jitter = 30,
    } = options;

    // TODO: Check cache
    const cached = await this.redisService.get(key);

    if (cached) {
      return JSON.parse(cached) as T;
    }

    const lockKey = `lock:${key}`;
    const lockToken = randomUUID();

    // TODO: Try to acquire lock
    const lockAcquired = await this.redisService.acquireLock(
      lockKey,
      lockToken,
      lockTtl,
    );

    if (lockAcquired) {
      try {
        // TODO: Double-check cache
        const cachedAfterLock = await this.redisService.get(key);

        if (cachedAfterLock) {
          return JSON.parse(cachedAfterLock) as T;
        }

        // TODO: Fetch from DB/API
        const data = await fetcher();

        // TODO: TTL jitter
        const finalTtl = ttl + Math.floor(Math.random() * jitter);

        // TODO: Store in cache
        await this.redisService.set(key, JSON.stringify(data), finalTtl);

        return data;
      } finally {
        // TODO: Release lock
        await this.redisService.releaseLock(lockKey, lockToken);
      }
    }

    // TODO: Another request is fetching the data
    for (let attempt = 0; attempt < retries; attempt++) {
      await this.sleep(retryDelay);

      const cachedAfterWait = await this.redisService.get(key);

      if (cachedAfterWait) {
        return JSON.parse(cachedAfterWait) as T;
      }
    }

    return fetcher();
  }

  private async sleep(ms: number): Promise<void> {
    await new Promise((resolve) => setTimeout(resolve, ms));
  }
}
